File: 1049-Last-Stone-WeightII/1049.go
func lastStoneWeightII(stones []int) int {
	sum := 0
	for _, v := range stones {
		sum += v
	}
	m := sum / 2
	dp := make([]bool, m+1)
	dp[0] = true
	for _, weight := range stones {
		for j := m; j >= weight; j-- {
			dp[j] = dp[j] || dp[j-weight]
		}
	}
	for j := m; ; j-- {
		if dp[j] {
			return sum - 2*j
		}
	}
}

File: 1001-1100/1049-Last-Stone-WeightII/1049.java
class Solution {
  public int lastStoneWeightII(int[] stones) {
    int sum = 0;
    for (int weight : stones) {
      sum += weight;
    }
    int m = sum / 2;
    boolean[] dp = new boolean[m + 1];
    dp[0] = true;
    for (int weight : stones) {
      for (int j = m; j >= weight; --j) {
        dp[j] = dp[j] || dp[j - weight];
      }
    }
    for (int j = m;; --j) {
      if (dp[j]) {
        return sum - 2 * j;
      }
    }
  }
}